### PR TITLE
Extend Notifications API and return pinned notifications by default

### LINKED ISSUE #12152: Notification API bugs.
- Gitea version (or commit ref): 1.13.0+dev-218-gec40e593c
- Git version: N/A
- Operating system: N/A
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [ ] SQLite
  - [x] N/A
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [ ] Not relevant
  - [x] Tried at [gitea.com](https://gitea.com/)
- Log gist:

## Description

There are two major bugs when trying to load a selection of notifcations [using the API](https://gitea.com/api/swagger#/notification/notifyGetList). These bugs were already confirmed by @6543.

### 1. Bug
From what I've seen, any request with ``all``=true leads to an internal server error (Response code: 500).

### 2. Bug
Any pinned notifications are not included in the results.

### PR DESCRIPTION
**Pull request by zeripath**
This PR extends the notifications API to allow specific notification statuses to be searched for and to allow setting of notifications to statuses other than read.

By default unread and pinned statuses will be returned when querying for notifications - however pinned statuses will not be marked as read.

Close #12152 

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by opyale** [comment]
Will these changes be backported to 1.12?

**Comment by mmarif4u** [comment]
Thanks @zeripath for the fixes.

I hope these fixes makes it to 1.12.2.

**Review by lafriks — APPROVED** [review]

File: routers/api/v1/notify/repo.go [hunk 1]
```diff
@@ -135,20 +190,34 @@ func ReadRepoNotifications(ctx *context.APIContext) {
...
+		statuses := ctx.QueryStrings("status-types")
+		opts.Status = statusStringsToNotificationStatuses(statuses, []string{"unread"})
+		log.Error("%v", opts.Status)
 	}
 	nl, err := models.GetNotifications(opts)
 	if err != nil {
 		ctx.InternalServerError(err)
 		return
 	}
+	for _, t := range nl {
```
Comment (lafriks) [inline routers/api/v1/notify/repo.go]: This looks like debug left

**Review by lafriks — CHANGES_REQUESTED** [review]

File: routers/api/v1/notify/repo.go [hunk 2]
```diff
@@ -135,20 +190,34 @@ func ReadRepoNotifications(ctx *context.APIContext) {
...
+		log.Error("found thread: %d %d", t.ID, t.Status)
+	}
```
Comment (zeripath) [inline routers/api/v1/notify/repo.go]:
```suggestion
```

**Review by lafriks — APPROVED** [review]

**Comment by lunny** [comment]
CI failed.

**Comment by zeripath** [comment]
CI failure was not related to this change because this change does not affect those endpoints - certainly there were issues because of unchanged tests earlier.

File: models/notification.go [hunk 3]
```diff
@@ -89,8 +89,14 @@ func (opts *FindNotificationOptions) ToCond() builder.Cond {
 	if opts.IssueID != 0 {
 		cond = cond.And(builder.Eq{"notification.issue_id": opts.IssueID})
 	}
-	if opts.Status != 0 {
-		cond = cond.And(builder.Eq{"notification.status": opts.Status})
+	if len(opts.Status) == 1 {
```
Comment (6543) [inline models/notification.go]: cant we just use an "IN" condition if Status != 0 ?

Comment (zeripath) [reply models/notification.go]: Yes, but there are only 3 status types in total.

Comment (6543) [reply models/notification.go]: yes so it is fine for sqlite too - I dont see why we add go complexytiy what can be reached with a simple SQL statement - but thats only my point of view :)

Comment (zeripath) [reply models/notification.go]: done

File: models/notification.go [hunk 4]
```diff
@@ -89,8 +89,10 @@ func (opts *FindNotificationOptions) ToCond() builder.Cond {
...
```
Comment (6543) [inline models/notification.go]: I think L95 handle L93 too

File: models/notification.go [hunk 5]
```diff
@@ -89,8 +89,10 @@ func (opts *FindNotificationOptions) ToCond() builder.Cond {
...
+		cond = cond.And(builder.Eq{"notification.status": opts.Status[0]})
+	} else if len(opts.Status) > 1 {
+		cond = cond.And(builder.In("notifiation.status", opts.Status))
```
Comment (zeripath) [inline models/notification.go]:
```suggestion
		cond = cond.And(builder.In("notification.status", opts.Status))
```

File: models/notification.go (hunk 4 above)
Comment (6543) [inline models/notification.go]:
```suggestion
```

File: models/notification.go [hunk 6]
```diff
@@ -89,8 +89,10 @@ func (opts *FindNotificationOptions) ToCond() builder.Cond {
 	if opts.IssueID != 0 {
 		cond = cond.And(builder.Eq{"notification.issue_id": opts.IssueID})
 	}
-	if opts.Status != 0 {
-		cond = cond.And(builder.Eq{"notification.status": opts.Status})
+	if len(opts.Status) == 1 {
+		cond = cond.And(builder.Eq{"notification.status": opts.Status[0]})
```
Comment (6543) [inline models/notification.go]:
```suggestion
```

File: models/notification.go [hunk 7]
```diff
@@ -89,8 +89,10 @@ func (opts *FindNotificationOptions) ToCond() builder.Cond {
...
+	} else if len(opts.Status) > 1 {
```
Comment (6543) [inline models/notification.go]:
```suggestion
	if len(opts.Status) != 0 {
```

**Review by 6543 — CHANGES_REQUESTED** [review]

**Comment by zeripath** [comment]
I'm not certain that using IN for 1 argument is actually worth it - but then again it's arguably a premature optimisation to do the len() check - so in which case I will do it.

**Review by 6543 — APPROVED** [review]

**Comment by lafriks** [comment]
:rocket:

**Comment by 6543** [comment]
should we partialy backport it to 1.12? (so that pinned ones are with the unreaded) ?

**Comment by zeripath** [comment]
I guess we can claim it is a bug.

**Pull request merged**